### ISSUE TITLE
[Marabou With Gurobi] How can I debug my model if Gurobi says it's infeasible 

### ISSUE THREAD

**Issue by HatsuneMK00** [NeuralNetworkVerification/Marabou#551]
I'm using Marabou with Gurobi to do LP relaxation stuff. It seems like my code has some bugs and Gurobi gives a log saying 'model is infeasible'. I'm wondering whether there is a way to find out the reason why the model is infeasible, like what kind of log information may help and which verbosity level should I set? Thanks a lot!
(I'm using Marabou via its python API so I think I don't have direct access to Gurobi which means I need to debug my model with Marabou.)

**Comment by anwu1219** [comment]
Hi @HatsuneMK00 , is the issue related to the one you mentioned in #547 ?
Could you please check if you can dump the query with this method below before calling the net.solve() method?
https://github.com/NeuralNetworkVerification/Marabou/blob/76b8eaf23518ca468c2cf05b742e3b4c858a64c3/maraboupy/MarabouNetwork.py#L372-L379

For example: net.saveQuery("failedQuery.txt")

If you could share this file with me I can take a quick look.

**Comment by HatsuneMK00** [comment]
Yes this is related to the issue #547 . And I think it's about debugging the model using Marabou with Gurobi so I opened a new issue. 
And again thanks a lot for your kind help! It seems like that i have some conflicts on the constraints and the upper&lower bound I set to inputVars, and I think I've solved it by now.
Therefore, I'm wondering if I can control some settings of Gurobi like number of worker process or timeout within Marabou?

**Comment by anwu1219** [comment]
Hi @HatsuneMK00 , yeah you can control those, by setting the numWorkers and timeoutInSeconds parameters in the option object (I presume the solveWithMILP flag is already set to True?). Those also apply to the Gurobi mode.

https://github.com/NeuralNetworkVerification/Marabou/blob/69dc5eb71b5f4e7a813f20a4e1a01b8387102258/maraboupy/Marabou.py#L121-L128

**Comment by HatsuneMK00** [comment]
Thank you so much! I've tried it out but it seems not works for me?
I call the `network.solve()` with the following options:
```python
options = Marabou.createOptions(numWorkers=32, timeoutInSeconds=3600, solveWithMILP=True)
```
And the output in terminal is still like this:
<img width="704" alt="截屏2022-03-25 16 07 28" src="https://user-images.githubusercontent.com/32505133/160080970-35e60fbe-e6c0-4ca4-8209-62a3f815ec17.png">

**Comment by anwu1219** [comment]
The options looks good to me.
Could you please make sure you've pulled from the latest master? This functionality was enabled quite recently.

**Comment by HatsuneMK00** [comment]
It seems to work after I update to the latest master. But error occurred in cxxtest when I run `cmake --build .` I can still run python API and the Gurobi output says I'm using up to 32 threads, but the cpu usage is still 100% while I am expecting it to go over 100%. So I'm not sure whether the settings is working. Please tell me if there is something wrong. Thank you so much!
The cmake option I'm using is like this:
```shell
cmake .. -DENABLE_GUROBI=ON -DGUROBI_DIR=/home/makise/gurobi951/linux64 -DBUILD_PYTHON=ON -DOPENBLAS_DIR=/home/makise/OpenBLAS
cmake --build .
```
And also I have dumped the cmake build log using tee. The stderr.log contains the error message given by cmake. The image is identical with the stderr.log. 
<img width="1055" alt="image" src="https://user-images.githubusercontent.com/32505133/160242710-b1fa38ab-0717-4ebe-a24f-ba75dbf7fd4e.png">
[stdout.log](https://github.com/NeuralNetworkVerification/Marabou/files/8355559/stdout.log)
[stderr.log](https://github.com/NeuralNetworkVerification/Marabou/files/8355560/stderr.log)

**Comment by HatsuneMK00** [comment]
And by the way, I'm curious on the time complexity of verification since I've got some query that seems to run forever. So I think knowing the (approximately) time complexity could help me figure out whether there is something wrong with my query.

**Comment by anwu1219** [comment]
Hum.. the build error seems to be in the regression test of a new module we added recently. I think it doesn't affect the usage (since that module is not put into used) but we will fix that soon. 
If the Gurobi output suggests it's using up-to 32 threads, I think that means the parameter is correctly set in Gurobi. Gurobi might not decide to use-up all the threads due to the parallelization scheme it's using.
Regarding complexity. For piecewise-linear constraints, the verification problem is NP-complete. If there are a large number of activation functions or disjunctions, the complete procedure can become quite slow. We do run abstraction-based methods before the complete procedure to prune the search space or (potentially) prove the property. 
If you have hard instance that our solver is having a hard time processing, I'd be curious to take a look.

**Comment by HatsuneMK00** [comment]
Thank you very much for your kind reply. After I posted my comment, I have read your previous work named "reluplex", trying to learn more about the verification method. I guess it's my fault to put way too much constraints on the model, so I'm trying to reduce the number of constraints.
